### PR TITLE
gh-1483 - Extending the GetWalks operations to include Entities

### PR DESCRIPTION
**Pull request by t616178**

### CODE REVIEW

**Comment by t616178** [comment]
Closing to make further additions.

**Comment by t616178** [comment]
Reopened after adding in full implementation. This could probably do with a bit of cleaning up - any suggestions would be appreciated.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/graph/adjacency/AdjacencyMap.java [hunk 1]
```diff
@@ -23,31 +23,35 @@
...
  * An {@code AdjacencyMap} is used to store the contents of a graph in memory in
  * a format which can easily be interrogated.
  *
- * @param <V> the type of object representing the vertices
- * @param <E> the type of object representing the edges
+ * @param <V>  the type of object representing the vertices
+ * @param <ED> the type of object representing the edges
  */
-public class AdjacencyMap<V, E> {
+public class AdjacencyMap<V, ED> {
```
Comment (p013570) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/graph/adjacency/AdjacencyMap.java]: Do we need the ED generic type? Would it ever be anything other than an Edge?

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/graph/GraphWindow.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * A GraphWindow can be used to create a graph representation which tracks changes to
+ * a graph, for example temporal changes or different views on a parent graph seen
+ * during a Walk.
+ *
+ * @param <V> the type used to represent vertices
+ * @param <EN> the type used to represent entities
+ * @param <ED> the type used to represent edges
+ */
+public class GraphWindow<V, EN, ED> {
```
Comment (p013570) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/graph/GraphWindow.java]: Do we need these generic types?

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/graph/entity/EntityMap.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * An {@code EntityMap} is used to store the contents of a graph in memory in a
+ * format which can easily be interrogated.
+ *
+ * @param <V>  the type of object representing the vertices
+ * @param <EN> the type of object representing the entities
+ */
+public class EntityMap<V, EN> {
```
Comment (p013570) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/graph/entity/EntityMap.java]: Similarly, do we need the generic types?

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/data/WalkDefinition.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+import uk.gov.gchq.gaffer.operation.impl.get.GetElements;
+import uk.gov.gchq.gaffer.operation.io.Input;
+import uk.gov.gchq.gaffer.operation.io.Output;
+
+import java.util.List;
+
+/**
+ * A {@code WalkDefinition} describes how to carry out a single step in a {@link uk.gov.gchq.gaffer.data.graph.Walk}.
+ */
+public class WalkDefinition implements Cloneable {
```
Comment (p013570) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/data/WalkDefinition.java]: Perhaps reorder the fields to: preFilters, operation, postFilters. Or add this annotation to improve the serialisation:
@JsonPropertyOrder({ "preFilters", "operation", "postFilters"})

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/data/WalkDefinition.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+     *
+     * @param preFilters  the preFilter operation chain
+     * @param postFilters the postFilter operation chain
+     * @param operation   the GetElements operation
+     */
+    @JsonCreator
+    public WalkDefinition(@JsonProperty("preFilters") final OperationChain<? extends Iterable<ElementId>> preFilters,
+                          @JsonProperty("postFilters") final OperationChain<? extends Iterable<ElementId>> postFilters,
+                          @JsonProperty("operation") final GetElements operation) {
+        this.preFilters = preFilters;
```
Comment (p013570) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/data/WalkDefinition.java]: This allows pre and post filters to be null. Throughout the rest of this class you will need to have null checks for the fields.

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/GetWalks.java [hunk 6]
```diff
@@ -47,10 +49,9 @@
  */
 public class GetWalks implements
         InputOutput<Iterable<? extends EntityId>, Iterable<Walk>>,
-        MultiInput<EntityId>,
-        Operations<GetElements> {
+        MultiInput<EntityId> {
 
-    private List<GetElements> operations;
+    private List<WalkDefinition> walkDefinitions;
```
Comment (p013570) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/GetWalks.java]: We need to come up with a way to get this to work with GraphHooks and the Graph. The Graph needs to be able to update the views on all nested operations. GraphHooks also need to be able to iterate around all operations and update them if required.

**Comment by p013570** [comment]
Using the Road Traffic example the following causes an error:

```json
{
   "class": "uk.gov.gchq.gaffer.operation.impl.GetWalks",
   "walkDefinitions": [
      {
         "preFilters": {
            "class": "uk.gov.gchq.gaffer.operation.OperationChain",
            "operations": []
         },
         "postFilters": {
            "class": "uk.gov.gchq.gaffer.operation.OperationChain",
            "operations": []
         },
         "operation": {
            "class": "uk.gov.gchq.gaffer.operation.impl.get.GetElements",
            "view": {
               "edges": {
                  "RoadUse": {}
               },
               "entities": {}
            }
         }
      }
   ],
   "resultsLimit": 10000,
   "input": [
      {
         "class": "uk.gov.gchq.gaffer.operation.data.EntitySeed",
         "vertex": "South West"
      }
   ]
}
```


Error: Index: 0, Size: 0
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
        at java.util.ArrayList.rangeCheck(ArrayList.java:653)
        at java.util.ArrayList.get(ArrayList.java:429)
        at uk.gov.gchq.gaffer.data.graph.adjacency.AdjacencyMaps.get(AdjacencyMaps.java:50)

**Comment by p013570** [comment]
Closing while changes are made.

**Pull request closed without merging**